### DISCUSSION TITLE
Decoupling Quickstart & CLI

### DISCUSSION THREAD

**Discussion by Lewiscowles1986** [Budibase/budibase#1751] [Ideas]
Have you thought about keeping quickstart simple and quick, and introducing the CLI as a later option?

A quickstart repo, or download option with the three files people will want might be an idea for the casual interest. What do you think of this? It could even be a downloadable zip.

I was looking to have a play with this, and it seemed confusing that you spoke a lot about docker and docker-compose on the Getting started page, but then immediately and irresumably jumped to npm. https://docs.budibase.com/getting-started

The minimum requirements are docker and docker-compose, then a git repo, tarball or zip might be a very simple way for people to pull, and run a single command. Even better, github gives them a download zip option, so you could be hand-wavy about where those three files come from, link people to a zip, and continue from there.

There were no clear links I could find to https://github.com/Budibase/budibase/tree/master/packages/cli/src/hosting, from the quickstart docs page; which it turns out just downloads two files, makes a third (which also has a download option), and then calls docker-compose up.
https://docs.budibase.com/self-hosting/hosting-methods/docker-compose#installing-docker
> `docker-compose --env-file hosting.properties up`

## Retracing steps

1. I started at the homepage, having made a note to checkout budibase, I went straight to your website via duck duck go.
2. I clicked to the get started free link at the bottom of the page via the callout, then I had to scan for quick start on the right.
3. I reassuredly read needs docker and docker-compose. Great I thought, no need for a VM to isolate this thing from my computer.
4. I Became immediately confused why the next instructions involved npm for a docker and docker-compose setup
5. I searched npmjs for the package (missed the github link next to dark mode in header of main site. I still don't see it on that docs page)
6. Again could not see a link to GitHub from https://www.npmjs.com/package/@budibase/cli
6. Back to the getting started page, saw a wget link. It downloaded a 60MB binary, and I immediately started worrying about this software being... non-obvious.
7. Searched duck duck go for budibase github
8. found the org and this monorepo
9. found files
10. read files
11. found https://docs.budibase.com/self-hosting/hosting-methods/docker-compose#installing-docker from the files

I think the script is really quite neat. I recognised some of the tech from other nodejs work I've done, but it was a bit much to git clone a folder which should have three files, and then run a one-liner, to need to also install nodejs & npm. I think others may definitely have found this. After all you wouldn't have caught my attention if someone were not evangelizing your platform on dev.to. Perhaps I am very particular, but 🤷  I think the quickstart could be improved. I very much think splitting "install our nodejs cli package" should be some steps later than "see what we have made".

Even the docker, to be honest is less click-and-go than an OVA, or similar one-click-able thing. It's a low-code solution. I as a coder am already noticing envoy, couchdb, lots of things that if I'm honest are cool, you have a split application and worker. Very impressive. But also, decidedly coder heavy. Your graphic artist is going to struggle to get started with this.

* no minimum nodeJS version mentioned. I run nvm, but does everyone on the internet?
* obscured simplicity
* no machine recommended and minimum specs (I'm running on an i7 laptop, but it works in a docker-machine with 1GB RAM)

It took seconds to get setup using the three file method. I spent more time writing this and reading your source due to the NodeJS CLI option.

**Comment by mike12345567** [comment]
Hi @Lewiscowles1986 thank you for this great feedback and write-up, this must have taken some time to put together!

First off I think you're right that we should have multiple methods in our docs for setting up quickly and easily, there are definitely a few people that prefer just managing it themselves as its easier/faster if you understand Docker/Docker compose. I'll raise this with the team and get a "manual" section added to the docs which has the links the latest versions of the 3 required files. This should probably actually just link to our docs guide on setup via docker/docker compose that you found, this should be referenced very early in our getting started guide as an option.

A big point i'd initially like to make is that we recognise getting started/trying Budibase right now is quite painful, for those who don't have experience with these sorts of technologies it can be quite a pain and we want to reduce this, so that more people can find out how awesome Budibase is! We're in the process of creating a sandbox/cloud environment where people can trial a basic version of Budibase, then if they like it they can proceed with self hosting (exporting their app from the sandbox and continuing on with its development/deployment). For actual deployments we do feel the CLI is a very useful utility to have as right now it can run updates/upgrades and in the future it will be used for migrations and other infrastructure related tasks.

Also, if you have some examples of the click-and-go OVA solutions you've seen in other similar projects, I would be interested to see this, as if we can offer other easier solutions that would be great.

As for minimum node versions, I believe I added a check to the codebase to warn the user if they have not got a new enough version installed, however I believe the minimum version required by it is quite an old version (can't quite remember off the top of my head!) We included the built binary as an option for people that don't have Node/NPM installed as this is just bundled together with [pkg](https://github.com/vercel/pkg) although maybe the docs don't make it clear that is the purpose of the downloadable binary.

Minimum specs of the cluster is something that we need to address, I will bring this up with the rest of the team and we'll look at getting something about this in the docs, although right now the requirements are pretty minimal, you can run the whole thing out of pretty small digital ocean droplets with no issues.

Just to add, I'm the engineer that put together the CLI originally, we plan to do a lot more with the CLI in the future - we want it to be a useful utility for developers to be able to extend and use Budibase. For example data source integrations, backend automations, various other parts of the builder, will be extensible and we want developers to be able to use environments they already know/use, then they can use the Budibase CLI to build/integrate painlessly. This is all future work but just wanted to give some context for the CLI's existence as right now it is somewhat light on functionality compared to what we hope it will do.

**Reply by Lewiscowles1986** [reply]
Take notes devs. This is how to respond. It's hard to get anything to everyone liking, but how you've handled this is very product-focused, very mature and I'm further impressed.  This is more than I imagined you would have to offer on this. Very cool indeed. ❤️ 

As for examples, bitnami were who introduced me to easy deploys over a decade ago offer them, and they have everything from low-level primitives like [nginx](https://bitnami.com/stack/nginx/virtual-machine); To higher-level applications like [WordPress](https://bitnami.com/stack/wordpress/virtual-machine), and more. To find out more about [bitnami stacks](https://bitnami.com/stacks), is definitely a good place to look. (they offer OVA files)

As far as I know, they often provide a poor upgrade pipeline, unlike something like a vagrant box (but that is too dev focused again, like docker). Really I guess OVA is yet another dev format.

I was thinking that the first install should not be someone production deployment, so Quick Start as a place to as you say, provide a sandbox. 

The real thing to solve for product people I know, would be able to natively run on all three platforms, and would grab it's own vendor dependencies, and keep outside of system PATH. Maybe Sandbox will help with that audience. 

The key is to give enough of a preview, to let someone try it out with the minimum fuss or effort, that they ask a dev to look for more than a few seconds.

I mention OVA because after a single install (VirtualBox), they can download the file, run it and be presented with a GUI in an OS you as the package-maintainer control; so no need for the user to type in commands, install npm or docker.

I've since found your YouTube playists too https://www.youtube.com/channel/UCsPsmlrhbogYR3_JDJym1fg/playlists
Although fast-paced, it's very low-fuss and would have pointed me at exactly what I needed as a dev.

If you are looking to iterate on CLI, is there a place people can see the roadmap? It's AGPL, so maybe people can contribute too?

You did a lot better than `posthog` on things I wanted to look at this weekend, so for a dev it's not the end of the world, just a little confusing.

**Reply by shogunpurple** [reply]
Hey @Lewiscowles1986, this is awesome feedback. Bitnami are a great example and I've been a big fan of their offerings on the AWS marketplace and more for quite a while. We had a discussion earlier about some of the things that you mentioned - we will be updating the docs with a docker quickstart to be more in line with how most self-hosted docker based software is provided by other open source projects.

The roadmap is another timely point. We have been planning our Q3 roadmap this week, and have a github project for our public roadmap. This is slightly outdated at this time, but will be brought completely up to date over the next day or two - with all the big initiatives we plan to ship over the next couple of months. You can find that roadmap here:
https://github.com/Budibase/budibase/projects/10

If contributing is something that you would be interested in, we welcome all contributors. You can get involved in github discussions, have a look at some of our github issues, and even join our discord server at the following link to chat development/contribution with the team:

https://discord.gg/r9CnWuA4

With regards to providing an OVA and the general virtual machine route, it's not something that we could achieve without quite a bit of effort and changes to the underlying architecture, given that we are very heavily dependent on a docker based setup. We would look at this if the demand is there in future. Other docker-based deploy targets we are looking at supporting for self hosting in the nearer term are kubernetes, terraform and AWS cloudformation to manage ECS. 

Also interesting you mentioned our friends over at posthog! We use their product internally and they've done a fantastic job with it. Be sure to check it out if you ever get a chance!
